functionalization <> LTC integration

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80043

I made a new PR for https://github.com/pytorch/pytorch/pull/75527.

It looks like there's a bug in the windows CI scripts that was causing
flaky failures, that disappear when I create a new PR. example failure:
https://github.com/pytorch/pytorch/runs/6999272635?check_suite_focus=true

bdhirsh committed:
new PR for https://github.com/pytorch/pytorch/pull/75527.

It looks like there's a bug in the windows CI scripts that was causing
flaky failures, that disappear when I create a new PR. example failure:
https://github.com/pytorch/pytorch/runs/6999272635?check_suite_focus=true

[ghstack-poisoned]

bdhirsh commented:
Cc @wconstab @Krovatkin There we’re some strange windows only failures that went away when I made a new PR- do you mind giving a stamp?

wconstab commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -260,6 +263,49 @@ const char* FunctionalTensorWrapper::tensorimpl_type_name() const {
+      // with "allow_tensor_metadata_changes" set to false.
+      // We *always* want this flag to be true for FunctionalTensorWrapper though,
+      // Since mutations later on can changes the size/stride metadata on the wrapper.
+      /*allow_tensor_metadata_change=*/true);
```
I don't know enough about this to comment, but did someone else weigh in?  Do you have a test case that motivates setting allow_ to true here?

bdhirsh replied:
Alban did previously - I actually ended up fixing this (just updated the PR). Comment here: https://github.com/pytorch/pytorch/pull/75527#discussion_r905697689

wconstab commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -552,5 +619,93 @@ void setFunctionalizationReapplyViewsTLS(bool reapply_views) {
+    }
+  }
+
+  {
```
this function looks right to me. thanks for explaining the key set manipulation :)

wconstab commented on aten/src/ATen/FunctionalTensorWrapper.h:
```diff
@@ -251,5 +272,38 @@ class TORCH_API FunctionalizationReapplyViewsGuard {
+        c10::KernelFunction::make_boxed_function<functionalize_op_helper>,
+        nullptr,
+        op,
+        c10::DispatchKeySet(), // we know that the cpu_fallback doesn't use the
```
what does this mean?  any significance for ptxla?

wconstab commented on aten/src/ATen/native/ts_native_functions.yaml:
```diff
@@ -141,42 +141,65 @@ full_codegen:
   - upsample_nearest2d_backward
   - zero.functional
   - narrow_copy.SymInt
+  - alias_copy
```
i guess you went through and added support for the superset of supportable ops, even if they weren't yet supported?  Is there test coverage for these?  I assume for ops like `split_copy` you attempted to add it but then it failed something so leaving it for later?

bdhirsh replied:
Yep - I'll update with a quick comment on that. I tried to add them originally, but it looked like the LTC codegen didn't support ops with `TensorList` outputs.

wconstab commented on torch/csrc/lazy/core/tensor.h:
```diff
@@ -242,6 +242,36 @@ TORCH_API LazyTensorPtr GetLtcTensorOrCreateForWrappedNumber(
 TORCH_API at::Tensor CreateAtenFromLtcTensor(const LazyTensorPtr& ltc_tensor);
 TORCH_API at::Tensor CreateAtenFromLtcTensor(LazyTensor&& ltc_tensor);
 
+// Given a non-lazy tensor, creates a lazy tensor on the specified (lazy)
```
guessing this should be tagged with `Note [Lazy Tensor Functionalization]`

wconstab commented on torch/csrc/lazy/ts_backend/ts_native_functions.cpp:
```diff
@@ -525,124 +469,160 @@ at::Tensor& LazyNativeFunctions::normal_(
+    at::IntArrayRef input_sizes,
+    int64_t dim,
+    int64_t index) {
+  return at::functionalization::functionalize_aten_op<ATEN_OP(
```
for some reason i thought these ops were being handled by the codegen.  it looks like they could be, but maybe that was just left for future work?

wconstab commented on torch/csrc/lazy/ts_backend/ts_native_functions.cpp:
```diff
@@ -525,124 +469,160 @@ at::Tensor& LazyNativeFunctions::normal_(
+  auto out_wrapped = at::functionalization::impl::to_functional_tensor(out);
+  // directly call the composite kernel from core.
+  // Make sure to re-enable functionalization first.
+  auto curr_tls = c10::impl::tls_local_dispatch_key_set();
```
could this block of key-set manipulations be wrapped up in some helper that returns a new 'guard' object?

bdhirsh replied:
I actually did that originally, and hit some windows-only symbol issues: https://github.com/pytorch/pytorch/runs/6937794259?check_suite_focus=true.

I figured that making a new guard object was a bit of overkill anyway, because I think it shouldn't have multiple call-sites long term (only used in the implementation in `FunctionalTensorWrapper.cpp`). It's also used here in this file for the `logsumexp_out` op - but:
(1) that's only because of some weirdness with that out= op (it's an out= composite op that calls into view ops)
(2) Long term, it would probably ideal for LTC to only have to deal with functional variants of operators anyway

wconstab replied:
ok, not a big deal.

wconstab commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -486,6 +486,25 @@ def setUpClass(cls):
+        import torch._lazy.ts_backend
+        global lazy_ts_backend_init
+        if not lazy_ts_backend_init:
+            # Nead to connect the TS backend to lazy key before running tests
```
Need, haha

wconstab reviewed (commented):
ok this looks good to me overall. there were parts i didn't understand completely and also some nits.  I would be OK stamping, not sure if you want to get any other reviews first, or you've already gotten enough feedback from the first incarnation of this PR?

bdhirsh commented:
@wconstab I got a decent amount of review already, so I’ll take a stamp if you’re ok with it and leave some of the items as follow ups. (Im away from my laptop, but I’ll respond to the individual comments later).

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3635,12 +3636,12 @@
   dispatch:
     CPU: pixel_shuffle_cpu
-    CompositeExplicitAutograd: math_pixel_shuffle
+    CompositeExplicitAutogradNonFunctional: math_pixel_shuffle
```
> the point of this set of changes is to mark 'unsafe' kernels that do non-functional things

yep! The point is mostly a better error message - the `*NonFunctional` key will set things up so that this kernel does not get registered to the Lazy + XLA dispatch keys. We don't want that, because the kernel for this function decomposes into view operators.

> later that is utilized by e.g. LTC IR codegen to call into the functionalize helper

Pretty much. Well, the LTC codegen doesn't know about it - there are ~8-9 of these "problematic" decompositions in total, and I manually wrote the code to call into the functionalize helper for those ops in `ts_native_functions.cpp`. Using in codegen instead is definitely an option - I just didn't bother because the amount of boilerplate was pretty low, and I expect these kernels to change very rarely (given that there are only 8-9 "problematic decompositions" out of our 2000+ aten ops, it looks like we don't add these very often).

bdhirsh commented on aten/src/ATen/native/ts_native_functions.yaml:
```diff
@@ -141,42 +141,65 @@ full_codegen:
+  # Why? These operators are all CompositeExplicitAutograd, which mean that they run
+  # after functionalization,
+  # but their implementations call view operators (which we need to functionalize away).
+  - block_diag
```
yep. I actually made that change already on the XLA side ([here](https://github.com/pytorch/xla/pull/3646)). @antoniojkim will need to make similar changes on the MLIR side (we talked about it over slack last week).

bdhirsh commented on torch/csrc/lazy/core/tensor.cpp:
```diff
@@ -532,5 +535,27 @@ at::Tensor CreateAtenFromLtcTensor(LazyTensor&& ltc_tensor) {
+      eager_tensor, torch::lazy::atenDeviceToBackendDevice(device));
+  auto out = torch::lazy::CreateAtenFromLtcTensor(lazy_self);
+  if (functionalize_output) {
+    // See Note [Lazy Tensor Functionalization]
```
whoops thanks - meant to link this to the comment in `to_lazy_tensor`

bdhirsh commented on torchgen/gen_lazy_tensor.py:
```diff
@@ -162,6 +162,39 @@ def validate_shape_inference_header(
+# Some helper functions for the codegen.
+def get_ltc_helper_fns() -> str:
+    return """\
+at::Tensor to_meta(const at::Tensor& tensor) {
```
Yeah.. Originally I actually put them into a header file (where I exported them with `TORCH_API`), but I ran into some strange issues where XLA/MLIR weren't able to find the symbols.

I spent a bit of time debugging but ultimately just decided to dump them directly into the codegen output. We have some precedent for doing that in the codegen, like [here](https://github.com/pytorch/pytorch/blob/817eb94ff494e07e94a9effb626ce2121f226a54/torchgen/dest/register_dispatch_key.py#L141).

bdhirsh commented on torchgen/gen.py:
```diff
@@ -2281,6 +2282,24 @@ def gen_op_headers(
             )
         },
     )
+    view_copy_with_symint_pairs: List[Tuple[NativeFunction, NativeFunction]] = []
```
Thanks for asking - this is part of the "race" where more SymInt support to LTC was getting added while I was landing this PR, so I added just enough SymInt support for functionalization to get the tests to pass.

The functionalization codegen has existing logic to generate kernels for all view_copy ops, that basically just call the underlying view op and then perform a clone.

That's no longer what we want for the SymInt overloads, though (cc @Krovatkin): we have an existing implementation for `narrow_copy.SymInt` ([link](https://github.com/pytorch/pytorch/blob/817eb94ff494e07e94a9effb626ce2121f226a54/aten/src/ATen/native/TensorShape.cpp#L930)), that just "concretizes" the sym ints and calls into the normal `view_copy` kernel.

Since we need `SymInt` overloads of all of our view copy operators anyway, I added codegen here to automatically generate those kernels for all `view_copy.SymInt` overloads: they will just call `.expectInt()` or `.toIntArrayRefSlow()` on all the sym int inputs, and call the non-symint overload.

To do that though, the codegen needs to know how to pair up the `view_copy` and `view_copy.SymInt` `NativeFunction` objects. That pairing up is what this code is doing.

I remember hearing from our dynamic shapes chat that in the end state we'll probably end up with just having a single `SymInt` overload for these ops instead of having both the symbolic + concrete overloads. I think that it'll be pretty easy to remove this codegen if/once we get there.

wconstab reviewed (approved):
thanks for answering my questions. Here's your stamp, and i resolved a lot of the questions but i left ones that might have a code change or comment change associated with them.

bdhirsh committed:
Update on "functionalization <> LTC integration"


I made a new PR for https://github.com/pytorch/pytorch/pull/75527.

It looks like there's a bug in the windows CI scripts that was causing
flaky failures, that disappear when I create a new PR. example failure:
https://github.com/pytorch/pytorch/runs/6999272635?check_suite_focus=true

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot rebase

pytorchmergebot committed:
Update on "functionalization <> LTC integration"


I made a new PR for https://github.com/pytorch/pytorch/pull/75527.

It looks like there's a bug in the windows CI scripts that was causing
flaky failures, that disappear when I create a new PR. example failure:
https://github.com/pytorch/pytorch/runs/6999272635?check_suite_focus=true

[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization <> LTC integration"


I made a new PR for https://github.com/pytorch/pytorch/pull/75527.

It looks like there's a bug in the windows CI scripts that was causing
flaky failures, that disappear when I create a new PR. example failure:
https://github.com/pytorch/pytorch/runs/6999272635?check_suite_focus=true

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge

antoniojkim commented:
Not sure how the merge process works, but I'm seeing a bunch of merge conflicts when I try to merge your branch into latest `master`

bdhirsh committed:
Update on "functionalization <> LTC integration"


I made a new PR for https://github.com/pytorch/pytorch/pull/75527.

It looks like there's a bug in the windows CI scripts that was causing
flaky failures, that disappear when I create a new PR. example failure:
https://github.com/pytorch/pytorch/runs/6999272635?check_suite_focus=true

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

antoniojkim commented:
> that's strange, since no merge conflicts are showing up in the CI. I just rebased locally on master and re-pushed again just to be safe though (although I didn't hit any merge conflicts locally either).

Yeah its weird. Your rebase appears to have worked though. I'm no longer seeing any merge conflicts locally. Thanks!

bdhirsh commented:
@janeyx99 - I think the `pytorch_env_restore.bat` file used in windows CI might have a bug? I have windows CI failing on this PR because that script errors out ([here](https://github.com/pytorch/pytorch/runs/7047177607?check_suite_focus=true)) - it looks like it can't parse some characters or something.


If you look at my PR description, I actually re-made this PR from a previous one because of similar errors ([linked](https://github.com/pytorch/pytorch/runs/6999272635?check_suite_focus=true) in the description) - that error looked like this:
```
C:\actions-runner\_work\pytorch\pytorch>if exist "C:/actions-runner/_work/pytorch/pytorch/build/win_tmp/ci_scripts/pytorch_env_restore.bat" (
call C:/actions-runner/_work/pytorch/pytorch/build/win_tmp/ci_scripts/pytorch_env_restore.bat  
 exit /b 0 
) 
Environment variable + 1816329f100a3e63d82889586085f8d46e587c38 make functionalization work better with subclasses not defined
Environment variable + 42480641196985f053591c4b184c089439e9b45f code-generate non-aliasing {view}_copy kernels not defined
Environment variable + f90bad13e4ffd77e607729ecabf3a1fd871f9fae register meta allocator not defined
```

It looks like somehow the CI is trying to dump my commit messages into a script and run it. FWIW, this looks like the same instance of that problem I had last week, where the `pytorch_env_restore.bat` file was somehow creating a file called `nonfunctional`, which failed windows CI.

Does that look flaky enough for me to go ahead and land?

bdhirsh commented:
(cc @pytorch/pytorch-dev-infra)

janeyx99 commented:
Ah it does look like a parsing error on <

janeyx99 commented:
Alright so here's my guess of what's going on:

In order to know what tests to not disable, we look for a "fixes #issue_num" in commit messages and we store those in an environment variable. Creating the restore_env script is essentially copying the env vars of the CI and putting them into a script so that your environment will be the same.

Since a PR can have multiple commit messages, the script is trying to copy all of those over and their multi-line-ness added with the fact that your commit messages had illegal batch characters in them (the < in this case) caused this issue when the script was called. 

What you can do to unblock:
- rename all your commit messages to avoid <, >, |, &, and ^ for the time being
- your CI should be green and then you can land

What I will try to do:
- Figure out how to get batch to copy over the multiline variable correctly
- Maybe just get rid of the new lines 
- Some other batch magic to fix this that is unknown to me at this moment

bdhirsh commented:
@janeyx99 sorry, I forgot to comment on the PR this morning - I actually... made a fresh PR [here](https://github.com/pytorch/pytorch/pull/80251), and got green CI so I went ahead and landing it. The windows failures are pretty wild but fixing them isn't an active blocker. I can go ahead and close this PR as a duplicate.
